src/checkers/inference/model/ConstraintManager.java

```diff
@@ -37,171 +30,186 @@
+     */
     public PreferenceConstraint createPreferenceConstraint(VariableSlot variable, ConstantSlot goal,
             int weight) {
-        if (variable == null || goal == null) {
-            ErrorReporter.errorAbort("Create preference constraint with null argument. Variable: "
-                    + variable + " Goal: " + goal);
-        }
-        return new PreferenceConstraint(variable, goal, weight, getCurrentLocation());
+        return PreferenceConstraint.create(variable, goal, weight, getCurrentLocation());
     }
 
+    /**
+     * Creates an {@link ExistentialConstraint} for the given slot and lists of constraints.
+     */
     public ExistentialConstraint createExistentialConstraint(Slot slot,
             List<Constraint> ifExistsConstraints, List<Constraint> ifNotExistsConstraints) {
-        // TODO: add null checking for argument.
-        return new ExistentialConstraint((VariableSlot) slot,
-                ifExistsConstraints, ifNotExistsConstraints, getCurrentLocation());
+        return ExistentialConstraint.create((VariableSlot) slot, ifExistsConstraints,
+                ifNotExistsConstraints, getCurrentLocation());
     }
 
+    // TODO: give location directly in Constraint.create() methods
     private AnnotationLocation getCurrentLocation() {
         if (visitorState.getPath() != null) {
-            return VariableAnnotator.treeToLocation(inferenceTypeFactory, visitorState.getPath()
-                    .getLeaf());
+            return VariableAnnotator.treeToLocation(inferenceTypeFactory,
+                    visitorState.getPath().getLeaf());
         } else {
             return AnnotationLocation.MISSING_LOCATION;
         }
     }
 
+    // All addXXXConstraint methods create a (possibly normalized) constraint for the given slots
+    // and issues errors for unsatisfiable constraints.
+
+    /**
+     * Creates and adds a {@link SubtypeConstraint} between the two slots to the constraint set,
+     * which may be normalized to {@link AlwaysTrueConstraint} or {@link EqualityConstraint}. An
+     * error is issued if the {@link SubtypeConstraint} is always unsatisfiable.
+     */
     public void addSubtypeConstraint(Slot subtype, Slot supertype) {
-        if ((subtype instanceof ConstantSlot)
-                && this.realQualHierarchy.getTopAnnotations().contains(((ConstantSlot) subtype).getValue())) {
-            this.addEqualityConstraint(supertype, (ConstantSlot) subtype);
-        } else if ((supertype instanceof ConstantSlot)
-                && this.realQualHierarchy.getBottomAnnotations().contains(
-                        ((ConstantSlot) supertype).getValue())) {
-            this.addEqualityConstraint(subtype, (ConstantSlot) supertype);
-        } else {
-            this.add(this.createSubtypeConstraint(subtype, supertype));
-        }
+        add(createSubtypeConstraint(subtype, supertype), SubtypeConstraint.class);
```

Comment:
For me, this line increases puzzles for me. `createSubtypeConstraint()` is not guaranteed to return `SubtypeConstraint`, then why make it inconsistent to always passing `SubtypeConstraint.class`? Seems adding inconsistencies here.

Reply:
If `createSubtypeConstraint()` gives back an `AlwaysFalseConstraint`, the error message shouldn't say `alwaysfalse.constraint.unsatisfiable`. It should say `subtype.constraint.unsatisfiable`. Hence.

Did you read the javadoc of add, or the discussion I started about tracking the non-normalized constraint?